plugin/pkg/scheduler/factory/factory.go

```diff
@@ -168,14 +169,41 @@ func (factory *ConfigFactory) createMinionLW() *cache.ListWatch {
 	}
 }
 
-// pollMinions lists all minions and returns an enumerator for cache.Poller.
+// pollMinions lists all minions and filter out unhealthy ones, then returns
+// an enumerator for cache.Poller.
 func (factory *ConfigFactory) pollMinions() (cache.Enumerator, error) {
-	list := &api.NodeList{}
-	err := factory.Client.Get().Resource("minions").Do().Into(list)
+	allNodes := &api.NodeList{}
+	err := factory.Client.Get().Resource("minions").Do().Into(allNodes)
 	if err != nil {
 		return nil, err
 	}
-	return &nodeEnumerator{list}, nil
+	nodes := &api.NodeList{
+		TypeMeta: allNodes.TypeMeta,
+		ListMeta: allNodes.ListMeta,
+	}
+	for _, node := range allNodes.Items {
+		if len(node.Status.Conditions) == 0 {
+			// If no condition is set, node health check is disabled (master flag
+			// healthCheckMinions is set to false), we add nodes unconditionally.
+			glog.Infof("No condition set for node %v", node.Name)
+			nodes.Items = append(nodes.Items, node)
+			continue
+		}
+		for _, condition := range node.Status.Conditions {
+			// TODO: Distinguish different conditions and implement better condition check.
+			switch condition.Kind {
+			case api.NodeReady:
+				if condition.Status == api.ConditionFull {
+					nodes.Items = append(nodes.Items, node)
+				} else {
+					glog.V(2).Infof("Ignore unhealthy node %v", node.Name)
+				}
+			default:
+				glog.Errorf("Unknown condition kind %v for node %v", condition.Kind, node.Name)
```

Comment:
> This way, we put NodeReady into a higher priority than NodeReachable, e.g. If NodeReady is True but NodeReachable is False...

Yeah, ready implies reachable, or it should. I'm also open to rejecting the node if it's either not ready or not reachable, if that makes you happier.

> BTW, does the revised code what you expect?

I don't like the code as written right now, mostly because it could behave differently depending on the order in which it examines statuses. (But also I'd use "condition" instead of "L" and I still don't think it should log on an unknown type of status.)